### PR TITLE
Stone walls & Mineable ore veins

### PR DESCRIPTION
**Pull request by KathrinBailey**
## About The Pull Request
• Adds a new mineral: stone.
• Adds a new wall, the stone wall.
• Adds generally-coded ore veins with support for multiple sprites & ore types.
• Adds stone brick as a construction material, with stacks.
• Adds veins for all ore except for bananium, titanium and uranium. Gonna have to do the last two at a later date.

As of now, all of this content remains adminspawn only.
image
image

## Why It's Good For The Game
This replaces the placeholder stone wall (or "dungeon wall") sprite with a completed sprite. It now also functions as a normal wall would, with complete auto-smoothing.
Ore veins function very similarly to Runescape ore.
If you mine Runescape ore, it 'depletes' the vein, which changes the sprite and makes it unminable. If anyone else is mining it, they won't get the ore. After a period of time, the ore vein will regenerate, change its sprite back and be minable again.
**Here are the following features ore veins hold:**
• Customizable mining time.
• Customizable amount of ore to drop.
• Customizable ore to drop, which can be replaced with anything in the game.
• Customizable ore regeneration/respawn times.
• Support for unlimited sprite variations, which are randomized whenever the ore vein spawns, the mapper does not have to manually place different types to break up the monotony.
• Toggle for if the sprite variations should be random or not.
With this being so general, you can effectively turn this into... a bullet ore vein if you wanted? You can mine 100 bullets over two minutes. You can mine a Durand if you really want.

## Changelog
:cl:
add: Stone walls with full auto-smoothing are now in the game.
/:cl:

### CODE REVIEW

**Comment by GoldenAlpharex** [comment]
> Mappers on their way to add SM shard and Captain's Spare mines :^)

File: modular_skyrat/modules/stone/code/stone.dm [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	icon_state = "wall-0"
+	base_icon_state = "wall"
+	sheet_type = /obj/item/stack/sheet/mineral/stone
+	slicing_duration = 150   //literal rock
+	explosion_block = 2
+	smoothing_flags = SMOOTH_BITMASK
+	smoothing_groups = list(SMOOTH_GROUP_CLOSED_TURFS, SMOOTH_GROUP_WALLS, SMOOTH_GROUP_STONE_WALLS)
+	canSmoothWith = list(SMOOTH_GROUP_STONE_WALLS)
+	custom_materials = list(/datum/material/stone = 4000)
+/turf/closed/indestructible/stone
```
Comment (Gandalf2k15) [inline modular_skyrat/modules/stone/code/stone.dm:58]:
```suggestion

/turf/closed/indestructible/stone
```

File: modular_skyrat/modules/stone/code/stone.dm [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+	icon = 'modular_skyrat/modules/stone/icons/wall.dmi'
+	icon_state = "wall-0"
+	base_icon_state = "wall"
+	mineral = /obj/item/stack/sheet/mineral/stone
+	walltype = /turf/closed/wall/mineral/stone
+	smoothing_flags = SMOOTH_BITMASK
+	smoothing_groups = list(SMOOTH_GROUP_CLOSED_TURFS, SMOOTH_GROUP_WALLS, SMOOTH_GROUP_STONE_WALLS)
+	canSmoothWith = list(SMOOTH_GROUP_STONE_WALLS)
+
+
```
Comment (Gandalf2k15) [inline modular_skyrat/modules/stone/code/stone.dm]:
```suggestion
```

File: modular_skyrat/modules/stone/code/stone.dm [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	merge_type = /obj/item/stack/ore/stone
+
+/turf/closed/wall/mineral/stone
+	name = "stone wall"
+	desc = "A wall made of solid stone brick."
+	icon = 'modular_skyrat/modules/stone/icons/wall.dmi'
+	icon_state = "wall-0"
+	base_icon_state = "wall"
+	sheet_type = /obj/item/stack/sheet/mineral/stone
+	slicing_duration = 150   //literal rock
```
Comment (Gandalf2k15) [inline modular_skyrat/modules/stone/code/stone.dm]:
```suggestion
	slicing_duration = 1.5 SECONDS  //literal rock
```

File: modular_skyrat/modules/stone/code/stone.dm [hunk 4]
```diff
@@ -0,0 +1,80 @@
+/obj/item/stack/sheet/mineral/stone
+	name = "stone"
+	desc = "Stone brick."
+	singular_name = "stone block"
+	icon = 'modular_skyrat/modules/stone/icons/ore.dmi'
+	icon_state = "sheet-stone"
+	inhand_icon_state = "sheet-metal"
+	mats_per_unit = list(/datum/material/iron=MINERAL_MATERIAL_AMOUNT)
```
Comment (Gandalf2k15) [inline modular_skyrat/modules/stone/code/stone.dm]:
```suggestion
	mats_per_unit = list(/datum/material/stone=MINERAL_MATERIAL_AMOUNT)
```

File: modular_skyrat/modules/stone/code/ore_veins.dm [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	if(W.use_tool(src, user, src.mining_time, volume=50))
+		to_chat(user, "<span class='notice'>You mine the [ore_descriptor].</span>")
+		if(ore_type && ore_amount && depleted == FALSE)
+			new ore_type(loc, ore_amount)
+		SSblackbox.record_feedback("tally", "pick_used_mining", 1, W.type)
+		// After giving	
+		depleted = TRUE
+		icon_state += "_depleted"
+		addtimer(CALLBACK(src, .proc/regenerate_ore), regeneration_time)
+//	After the ore vein finishes its wait, we make the ore 'respawn' and return the ore to its original post-Initialize() icon_state.
```
Comment (Gandalf2k15) [inline modular_skyrat/modules/stone/code/ore_veins.dm:69]:
```suggestion

//	After the ore vein finishes its wait, we make the ore 'respawn' and return the ore to its original post-Initialize() icon_state.
```

File: modular_skyrat/modules/stone/code/ore_veins.dm [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	/// How many unique sprites for ore we have, we will pick them at random.
+	var/unique_sprites = 1
+	/// If we should pick a random sprite for the ore vein or not.
+	var/random_sprite = TRUE
+
+/obj/structure/ore_vein/Initialize()
+	..()
+	if(random_sprite == TRUE)
+		icon_state += "[rand(1, (unique_sprites))]"
+	original_icon_state = icon_state
```
Comment (Gandalf2k15) [inline modular_skyrat/modules/stone/code/ore_veins.dm]:
```suggestion
```

You don't need to do this, check my comment further down.

File: modular_skyrat/modules/stone/code/ore_veins.dm [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	/// How much ore do we drop?
+	var/ore_amount = 1
+	/// If the ore vein has been recently mined. If so, we cannot mine and must wait for it to regenerate.
+	var/depleted = FALSE
+	/// How long it takes for the ore to 'respawn' after being mined.
+	var/regeneration_time = 15 SECONDS
+	/// How long it takes for a tool to mine the ore vein.
+	var/mining_time = 3 SECONDS
+	/// Our original icon_state to revert to upon regeneration.
+	var/original_icon_state = ""
```
Comment (Gandalf2k15) [inline modular_skyrat/modules/stone/code/ore_veins.dm:21]:
```suggestion
```
Ditto

File: modular_skyrat/modules/stone/code/ore_veins.dm [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+			new ore_type(loc, ore_amount)
+		SSblackbox.record_feedback("tally", "pick_used_mining", 1, W.type)
+		// After giving	
+		depleted = TRUE
+		icon_state += "_depleted"
+		addtimer(CALLBACK(src, .proc/regenerate_ore), regeneration_time)
+//	After the ore vein finishes its wait, we make the ore 'respawn' and return the ore to its original post-Initialize() icon_state.
+/obj/structure/ore_vein/proc/regenerate_ore()
+	depleted = FALSE
+	icon_state = original_icon_state
```
Comment (Gandalf2k15) [inline modular_skyrat/modules/stone/code/ore_veins.dm]:
```suggestion
	icon_state = initial(icon_state)
```

Comment (Gandalf2k15) [inline modular_skyrat/modules/stone/code/ore_veins.dm]: Also this should be done in update icon

**Review by Gandalf2k15 — APPROVED** [review]

**Pull request merged**